src/PrestaShopBundle/Security/SslMiddleware.php

```diff
@@ -0,0 +1,143 @@
+ * versions in the future. If you wish to customize PrestaShop for your
+ * needs please refer to https://devdocs.prestashop.com/ for more information.
+ *
+ * @author    PrestaShop SA and Contributors <contact@prestashop.com>
+ * @copyright Since 2007 PrestaShop SA and Contributors
+ * @license   https://opensource.org/licenses/OSL-3.0 Open Software License (OSL 3.0)
+ */
+
+declare(strict_types=1);
+
+namespace PrestaShopBundle\Security;
+
+use PrestaShop\PrestaShop\Core\ConfigurationInterface;
+use PrestaShopBundle\Controller\Api\OAuth2\AccessTokenController;
+use Symfony\Component\HttpFoundation\JsonResponse;
+use Symfony\Component\HttpFoundation\RedirectResponse;
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpFoundation\Response;
+use Symfony\Component\HttpKernel\Event\RequestEvent;
+
+/**
+ * Middleware that is triggered during `kernel.request` event on Symfony routing process, to redirect to HTTPS in some cases.
+ *
+ * If PS_SSL_ENABLED & (PS_SSL_ENABLED_EVERYWHERE | REFERER is HTTPS)
+ * Then redirect to the equivalent URL to HTTPS.
+ * If the request is an API call, we always redirect to HTTPS and to TLSv1.2+ if we detect a previous version
+ * Warning : will lose post data,
+ */
+class SslMiddleware
+{
+    private const AVAILABLE_SECURE_PROTOCOLS = ['tls/1.2', 'tls/1.3'];
+
+    /**
+     * @var ConfigurationInterface
+     */
+    private $configuration;
+
+    public function __construct(ConfigurationInterface $configuration)
+    {
+        $this->configuration = $configuration;
+    }
+
+    /**
+     * Registered as `kernel.request` event listener.
+     *
+     * If the condition needs a redirection to HTTPS, then the current process is interrupted, the headers are sent directly.
+     *
+     * @param RequestEvent $event
+     */
+    public function onKernelRequest(RequestEvent $event): void
+    {
+        // already SSL, do nothing more
+        if ($this->isSSLrequirementsMet($event->getRequest())) {
+            return;
+        }
+
+        if ($this->isApi($event->getRequest())) {
+            $this->upgradeProtocol($event);
+
+            return;
+        }
+
+        $enabled = (1 == $this->configuration->get('PS_SSL_ENABLED'));
+        $forced = (1 == $this->configuration->get('PS_SSL_ENABLED_EVERYWHERE'));
+        $serverParams = $event->getRequest()->server;
+        $refererSsl = ($serverParams->has('HTTP_REFERER') && strpos($serverParams->get('HTTP_REFERER'), 'https') === 0);
+
+        if ($enabled && ($forced || $refererSsl)) {
+            $this->redirectToSsl($event);
+        }
+    }
+
+    private function redirectToSsl(RequestEvent $event): void
+    {
+        $redirect = str_replace('http://', 'https://', $event->getRequest()->getUri());
+        $event->setResponse(new RedirectResponse($redirect));
+    }
+
+    private function isSSLrequirementsMet(Request $request): bool
+    {
+        if ($this->configuration->get('_PS_API_FORCE_TLS_VERSION_') === false) {
```

Comment:
This is used to get the value of the const `_PS_API_FORCE_TLS_VERSION_` but I think it's a really bad practice to use the configuration service to access this kind of configuration. It would be better to inject the value in the constructor of this middleware instead.

Though even this is not ideal so it may not be worth blocking the PR for this, actually the real correct way is blocked for now but it's related to our coming task of improving PrestaShop configuration which will likely rely on another service.

I think when we start improving this the feature that allows getting const from the Configuration service should be deprecated with trigger_error which will allow us to detect all this kind of usages, and then we'll be able to refactor all the places that rely on it.